feat: Add subtitle sync ui for easy alignment

armanckeser opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://jellyfin.org/docs/general/contributing/issues page.
-->

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->
* Chrome reactive change

https://github.com/user-attachments/assets/819fee3a-b33d-46ba-95c9-c0687a858a20

* Firefox before and after (pressed playback and stopped for the actual subtitles to re-render)

![Screenshot 2025-04-17 204703](https://github.com/user-attachments/assets/10c80899-eeb7-4da2-bca9-60cd1e376868)
![Screenshot 2025-04-17 204719](https://github.com/user-attachments/assets/78d42933-a544-4537-8c9c-5365810ba1f8)

* WebOS Browser

![webos-dev-tmp-674d30ba-3ec7-4d78-bad5-42802e6794d3](https://github.com/user-attachments/assets/08ba5925-7287-4cf7-8f26-2b5a436f8f73)


* Adds a dynamic subtitle sync rendering. Users can now stop where they would like a subtitle to start, align the time with the view and continue
* Adds api from the html player and playback manager to get the subtitles.

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->
#4364

dmitrylyzo reviewed (commented):
1. This only works with "custom subtitle element" (doesn't work in Chrome) and obviously only for text (SRT) subtitles.
2. We probably should update the subtitles timeline on `timeupdate`.
3. IMO, the timeline is too close to the offset slider, which makes you think they are tied.
4. Too long for mobile.

armanckeser commented:
Do you have recommendations for 1?

I am also open to some guidance on 2. I tried a bunch of things, but I feel like there is an efficient solution my lack of FE knowledge eludes me from.

dmitrylyzo commented:
> Do you have recommendations for 1?

When we use native API: https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/textTracks
I think for SSA/ASS and PGS we should hide the timeline.

> I am also open to some guidance on 2. I tried a bunch of things, but I feel like there is an efficient solution my lack of FE knowledge eludes me from.

Ideally, we should use [`PlaybackSubscriber`](https://github.com/jellyfin/jellyfin-web/blob/c9ccec834a759b0480a00fccdb2a2cdbe3954df4/src/apps/stable/features/playback/utils/playbackSubscriber.ts#L44).
But something like this could work: https://github.com/jellyfin/jellyfin-web/blob/c9ccec834a759b0480a00fccdb2a2cdbe3954df4/src/controllers/lyrics.js#L162-L189

armanckeser reviewed (commented):
Hmm, does this mean that platforms like WebOS wouldn't be able to use this implementation? I presume WebOS uses a chromiom browser.

armanckeser commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -103,6 +112,137 @@ function init(instance) {
+    events.forEach(event => {
+        // Convert ticks to seconds - these already have the currentPlayerOffset applied by the plugin
+        const startSec = event.StartPositionTicks / 10000000;
+        const endSec = event.EndPositionTicks / 10000000;
```
done

armanckeser commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -103,6 +112,137 @@ function init(instance) {
+
+        // Clean and add the text
+        const cleanText = stripHtmlTags(event.Text || '');
+        eventEl.textContent = cleanText;
```
removed

armanckeser commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -132,15 +273,27 @@ class SubtitleSync {
+
+                    // Initialize the timeline visualization with the current offset
+                    const currentOffset = playbackManager.getPlayerSubtitleOffset(player) || 0;
+                    renderSubtitleTimeline(currentOffset);
```
changed the entire structure

armanckeser commented:
Still have to think a bit about the realtime updates, the problem is getting a smooth timeline. Updating current implementation on every timeupdate results in the timeline moving choppily every second. I need to increase the resolution for the percentage calculations or add a css transition or something, but I am not sure yet

armanckeser commented:
I tried to implement a smooth scroll, but I just give up. I think this provides the value I would like it to provide. And even though the playback rendering is a bit janky, probably better than not having it at all.

armanckeser commented:
@dmitrylyzo thoughts?

Gauvino commented:
Absolutely incredible

armanckeser commented:
@dmitrylyzo Added screenshots from Firefox, Chrome, WebOS browser. Added realtime update ability and a video from Chrome showing that the changes are accurate with the realtime rendering. Changed the styling to better match Jellyfin Web style.

armanckeser commented:
@dmitrylyzo Would love to learn what types of tests or proof I could add to make this easier to review for the team.

dmitrylyzo reviewed (commented):
1. As before, I think the timeline is too close to the offset slider, which makes you think they are tied.
Close:
![timeline-old](https://github.com/user-attachments/assets/15b27704-3550-41b8-9b19-e1155e33baef)
Split:
![timeline](https://github.com/user-attachments/assets/851f016b-9ace-43b2-bd69-6060be091770)

2. My biggest concern right now is intermediate refactoring. I prefer to do refactoring first and then add the feature. Ideally, we should make commits atomic. If no one has any objections to the current commit sequence, so be it.

armanckeser commented:
>As before, I think the timeline is too close to the offset slider, which makes you think they are tied.

Ah, thanks for the clarification. I thought you meant that the current time itself was too close (which it was), but "split" looks much better

> My biggest concern right now is intermediate refactoring. I prefer to do refactoring first and then add the feature. Ideally, we should make commits atomic. If no one has any objections to the current commit sequence, so be it.

I see. Let me try a stacked approach, I am thinking a PR (or commit?) to refactor the original into OffsetController and the SubtitleSync object, and another one to add the timeline view. Does that sound good to you?

armanckeser commented:
A bit sad that a year has passed since I opened this PR. I don't think it should take over a year to add a nice feature

Lothyriel commented:
Can we review this please? This is a super nice feature

Commit:
Separate and refactor SubtitleSync and OffsetController

Extract offset state and interaction handling from SubtitleSync into a
dedicated OffsetController class. Use real private methods, a currentOffset
getter/setter, and a shared slider bubble helper. No user-facing behavior
changes. Paves the way for the subtitle sync timeline feature.

Add Armanc to CONTRIBUTORS.

Commit:
Add methods to expose current track events

Expose the current subtitle track events from the HTML video player and
the playback manager in a normalized seconds-based format. This allows the
subtitle sync UI to render a timeline of upcoming subtitle events.

Commit:
Add subtitle sync timeline for easy alignment

Render a scrolling timeline of subtitle events around the current playback
position below the offset slider, so users can visually align subtitles with
the video. Dragging the offset shifts the subtitle blocks against a fixed time
ruler, making the correct alignment easy to find.

The timeline reuses pooled DOM nodes and updates their positions on each player
time update, keeping it cheap enough to run during playback on mobile. It is
only shown for text subtitle tracks whose events can be read back from the
player.
